File: 54. Spiral Matrix/54.js
// 54. Spiral Matrix solution in JavaScript
// https://leetcode.com/problems/spiral-matrix/description/

/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
 var spiralOrder = function(matrix) {
    let result = [];

    let row = matrix.length;
    let column = matrix[0].length;
    let up = 0;
    let left = 0;
    let right = column - 1;
    let down = row - 1;

    while (result.length < row * column) {
        //To right
        for (let col = left; col <= right; col++) {
            result.push(matrix[up][col]);
        }
        //To down
        for (let row = up + 1; row <= down; row++) {
            result.push(matrix[row][right]);
        }
        //To left
        if (up != down) {
            for (let col = right - 1; col >= left; col--) {
                result.push(matrix[down][col]);
            }
        }
        //To top
        if (left != right) {
            for (let row = down -1; row > up; row--) {
                result.push(matrix[row][left]);
            }
        }

        left++;
        up++;
        right--;
        down--;
    }
    return result;
};

File: 54. Spiral Matrix/54.go
// 54. Spiral Matrix solution in GoLang
// https://leetcode.com/problems/spiral-matrix/description/

package main

func spiralOrder(matrix [][]int) []int {
	var result []int

	row := len(matrix)
	column := len(matrix[0])
	up := 0
	left := 0
	right := column - 1
	down := row - 1

	for len(result) < row*column {
		//To right
		for col := left; col <= right; col++ {
			result = append(result, matrix[up][col])
		}
		//To down
		for row := up + 1; row <= down; row++ {
			result = append(result, matrix[row][right])
		}
		//To left
		if up != down {
			for col := right - 1; col >= left; col-- {
				result = append(result, matrix[down][col])
			}
		}
		//To top
		if left != right {
			for row := down - 1; row > up; row-- {
				result = append(result, matrix[row][left])
			}
		}
		left++
		up++
		right--
		down--
	}

	return result
}
